### DISCUSSION TITLE
PCF8547 menu configuration not present in the web menu

### DISCUSSION THREAD

**Discussion by radudorurus** [arendst/Tasmota#18497] [Support request]
I wish to use ESP8266 with GPIO expander and I need some help since it does not work.

I succeeded this with an ESP32 devkit and custom compile for GPIO expander (one year ago). It worked but I choose to use the standard build because I had enough GPIOs on the ESP32.
Because the ESP32 cannot control an AC dimmer yet, I need to make this work on a 8266.
Also I wish to use wired connection with Wiz550 module if possible.
The device has some sort of safety level and I do not trust wifi (when you need it most, that's the moment it crashes, the device become unavailable and it cannot receive the command to power off all outputs).
But the wired LAN is the second step.

I bought for this application a board with 4 relays and an ESP-12F module. (the relays are controlled directly by jumpers tied to GPIO 12, 13 ,14 ,16.)
Additionally I have some PCF8574A i2C GPIO expanders.
I wanted to keep free the SPI GPIOs for the ethernet module.

I compiled the build with tasmocompiler from benzino77 using the following options .

8266 Generic
(#define USE_I2C                                // Add support for I2C
#define USE_PCF8574                            // [I2cDriver2] Enable PCF8574 I/O Expander (I2C addresses 0x20 - 0x26 and 0x39 - 0x3F) (+2k1 code)
//  #define USE_PCF8574_MODE2                    // Enable Mode2 virtual relays/buttons/switches (+2k3 code)
#define USE_PCF8574_SENSOR                   // Enable Mode1 inputs and outputs in SENSOR message (+0k2 code)
#define USE_PCF8574_DISPLAYINPUT             // Enable Mode1 inputs display in Web page (+0k2 code)
#define USE_PCF8574_MQTTINPUT                // Enable Mode1 MQTT message & rule process on input change detection : stat/%topic%/PCF8574_INP = {"Time":"2021-03-07T16:19:23+01:00","PCF8574-1_INP":{"D1":1}} (+0k5 code)
// #define USE_MCP230xx_ADDR 0x20)
OTA flash in 2 steps:
1. tasmota-minimal directly from web link
2. upload from file the custom build (web, homeassistant, rules, port expander, timers)

I set correctly the GPIO 4 and 5 for i2c.
I2C scan sees the expander address.
I can enable i2cdriver for specific device.
But the button option to configure the PCF8574 does not show up in the configuration menu as stated in the wiki.
https://tasmota.github.io/docs/PCF8574/#introduction

I  recompiled the bare minimum and without port expander (because the compiler ends most of the time with an error) and the result is the same. The configuration menu for PCF8574 does not show up.

Is there any console command to enable webinterface PCF8574 menu?
Does it worth trying to free the SPI ports for the ethernet module (I searched for somebody who did this and there is hard to find)?

**Comment by barbudor** [comment]
~~You choose Mode2 vs original mode which is what is described in the docs.
Mode2 is configured through a template like JSON stored in a pseudo rule.
Check how-to in the driver's source code~~

Sorry I read to fast. You seem to use Mode1
So everything should be fine with your compile options. 

Are your PCF detected with and `i2cscan`?

**Reply by barbudor** [reply]
And by the way, forget about Wiz550 it is not and will not be supported.
If you want Ethernet, get an ESP32

**Reply by radudorurus** [reply]
> And by the way, forget about Wiz550 it is not and will not be supported. If you want Ethernet, get an ESP32

I see these days the discussion about the AC dimming on ESP32.

So I have 2 variants:
1. Stick with ESP32 I already have. attach the wiz5500 to it and hope the AC dimming feature will be implemented
2. Go with 8266 which already have AC dimming and forget the ethernet connection.

**Reply by barbudor** [reply]
No
When I say forget about Wiz550, I mean TOTALLY forget about it, even on ESP32

Only standard PHY connected on MDIO are supported such as (but not limited to)
- https://templates.blakadder.com/wireless_tag_WT32-ETH01.html
- https://templates.blakadder.com/olimex_ESP32-POE.html
- https://templates.blakadder.com/lilygo_T-Internet-POE.html

For boards with relays you can use Dingtian boards:
- https://templates.blakadder.com/dingtian_DT-R008.html
- https://templates.blakadder.com/dingtian_DT-R016.html
- https://templates.blakadder.com/dingtian_DT-R032.html
==> Beware, those boards are now by default sold with a commercial firmware and ESP32 is protected against DYI reflashing.
For flashing Tasmota you must tell the vendor BEFORE placing the order so he will ship a board with Unprotected ESP32

**Reply by radudorurus** [reply]
> ~You choose Mode2 vs original mode which is what is described in the docs. Mode2 is configured through a template like JSON stored in a pseudo rule. Check how-to in the driver's source code~
> 
> Sorry I read to fast. You seem to use Mode1 So everything should be fine with your compile options.
> 
> Are your PCF detected with and `i2cscan`?

The PCF only i2c connected device was detected properly.

After one more try, the D1mini worked.
I cannot figure out why it did not work. 

The last try was:
reset device to defaults
flash from web
configure with backlog
flash over web interface the custom build
set the i2c GPIO.

I reproduced the steps without reset default on the 4x relay board and it is working now.
But the transistor driver has 1k in the base and 10k to GND. The PCF8574 outputs so low current that in the base of the transistor arrive only 0.6V, not enough to saturate it (the expander is powered from the 3.3V on the board which measured is 3.1V).

**Reply by barbudor** [reply]
PCF8574 are known to be weak current source. Due to the design of its pseudo-bidirectional port, it can sink a few mA when at LOW level but cannot source current at HIGH level.
Better to be used with MOSFET that Bipolar
You may have to modify your relay board

**Reply by radudorurus** [reply]
In that case...
I will put aside the ESP12F_4xRelay board for future projects.
Get rid of W5500 module.
Aquire ETH01 and hope the AC dimming feature will be implemented successfully.

I have one more project where I need some support. Probably I wil open another discussion for that.
A have an independent automation (for heating boiler and DHW/Buffer tanks) on ArduinoNano and I wanted to use an ESP01 as serial to MQTT bridge.
I read some info avout tasmota serial and the library needed to load on to arduino, but thats all.
The program memory available in the code is about 2-3kb. 
I know I can migrate everything on ESP but it is a lot of work and it cannot be done in Tasmota.
